src/__tests__/wait-for.js

```diff
@@ -255,3 +264,71 @@ test('the real timers => fake timers error shows the original stack trace when c
 
   expect((await waitForError).stack).not.toMatch(__dirname)
 })
+
+test('does not work after it resolves', async () => {
+  jest.useFakeTimers('modern')
+  let context = 'initial'
+  configure({
+    // @testing-library/react usage to ensure `IS_REACT_ACT_ENVIRONMENT` is set when acting.
+    unstable_advanceTimersWrapper: callback => {
+      const originalContext = context
+      context = 'act'
+      try {
+        const result = callback()
+        // eslint-disable-next-line jest/no-if
+        if (typeof result?.then === 'function') {
+          const thenable = result
+          return {
+            then: (resolve, reject) => {
+              thenable.then(
+                returnValue => {
+                  context = originalContext
+                  resolve(returnValue)
+                },
+                error => {
+                  context = originalContext
+                  reject(error)
+                },
+              )
+            },
+          }
+        } else {
+          context = originalContext
+          return undefined
+        }
+      } catch {
+        context = originalContext
+        return undefined
+      }
+    },
+    asyncWrapper: async callback => {
+      const originalContext = context
+      context = 'no-act'
+      try {
+        await callback()
+      } finally {
+        context = originalContext
+      }
+    },
+  })
+
+  let data = null
+  setTimeout(() => {
+    data = 'resolved'
+  }, 100)
+
+  await waitFor(
+    () => {
+      if (data === null) {
+        throw new Error('not found')
+      }
+    },
+    {interval: 50},
+  )
+
+  expect(context).toEqual('initial')
+
+  await Promise.resolve()
+
+  expect(context).toEqual('initial')
```

Comment:
Was `no-act` without the change. We had the following schedule-resolution order:


Without this change:
1. start `asyncWrapper`
1. start `unstable_advanceTimersWrapper`
1. leave `asyncWrapper`
1. leave `unstable_advanceTimersWrapper`

With this change:
1. start `asyncWrapper`
1. start `unstable_advanceTimersWrapper`
1. leave `unstable_advanceTimersWrapper`
1. leave `asyncWrapper`

In other words, before `asyncWrapper` and `unstable_advanceTimersWrapper` were interleaved instead of nested.